### PR TITLE
Add player HP decreasing

### LINKED ISSUE #24: Constantly decrease player health
- [ ] Create a component that requires `CombatEntity`
- [ ] Implement the feature that decreases the health
      Do **not** use `CombatEntity.TakeDamage` method because it is for combat only
- [ ] Attach the component to the player character

Maybe use coroutine?

```csharp
using Combat; // will probably need this due to namespacing
using UnityEngine;

[RequireComponent(typeof(CombatEntity))]
public class <ClassName> : MonoBehaviour
{
    // implementation
    // probably should use `GetComponent<CombatEntity>` and do stuff with it
}
```

### PR DESCRIPTION
**Pull request by ericsrna**

### CODE REVIEW

File: Assets/Scripts/Combat/CombatEntity.cs [hunk 1]
```diff
@@ -122,6 +122,11 @@ public void GetAttacked(int rawDamage)
             InternalCurrentHealth = currentHealth - (rawDamage - currentDefense);
         }
 
+        public void GetDecreased(int rawDamage)
```
Comment (JameeKim) [inline Assets/Scripts/Combat/CombatEntity.cs]: It could have a better name that describes exactly what this function does. For example, `DecreaseHealth`, `TakeRawDamage`, `TakeDamageWithoutDefense`, and so on. Its main goal is to decrease the health points, so the name should indicate that.

Another point here is that the name of this method starts with `Get` while it is not a getter and actually has side-effects. Generally, this is not advised.

File: Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs [hunk 2]
```diff
@@ -44,7 +55,15 @@ private IEnumerator DisableInputCoroutine(float seconds)
             disabled = true;
             yield return new WaitForSeconds(seconds);
             disabled = false;
-            currentCoroutine = null;
+            //currentCoroutine = null;
+            currentCoroutine = StartCoroutine(HealthDecreaseCoroutine(healthDecreasingTiming));
```
Comment (JameeKim) [inline Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs]: If you do this, it will be impossible to start this `DisableInputCoroutine` again because `DisableInputForSeconds` checks if there is a coroutine saved to the variable `currentCoroutine`. That variable was for keeping track of the `DisableInputCoroutine` to prevent multiple instnaces of that to be in action at the same time. Maybe the name of the variable was not clear. Sorry for that.

Also, you actually don't need to start `HealthDecreaseCoroutine` again because it is already done in the `Start` event function. This will actually cause two instances of that coroutine to be running, decreasing the health two times more.

File: Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs [hunk 3]
```diff
@@ -44,7 +55,15 @@ private IEnumerator DisableInputCoroutine(float seconds)
...
+        }
+
+        private IEnumerator HealthDecreaseCoroutine(float seconds)
+        {
+            combatEntity.GetDecreased(heathDecreasingValue);
+            yield return new WaitForSeconds(seconds);
+            currentCoroutine = StartCoroutine(HealthDecreaseCoroutine(healthDecreasingTiming));
```
Comment (JameeKim) [inline Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs]: Starting a new coroutine is a relatively expensive operation compared to simple looping and yielding from coroutines. It is advised to use a while loop with condition `true` to create an infinite loop.

For example,
```csharp
private IEnumerator SomeCoroutine(float interval)
{
    while (true)
    {
        yield return new WaitForSeconds(interval);
        DoSomething();
    }
}
```

Or for better performance, you can also cache the `WaitForSeconds` object.
```csharp
private IEnumerator SomeCoroutine(float interval)
{
    WaitForSeconds intervalYielder = new WaitForSeconds(interval);

    while (true)
    {
        yield return intervalYielder;
        DoSomething();
    }
}
```

**Review by JameeKim — CHANGES_REQUESTED** [review]
In the issue #24, there was a list of broken-up steps to act as a guideline. The first one suggested creating a new component for decreasing the health. The very reason I recommended so was to prevent all the complications that can happen when trying to do multiple things in a single file. It seems like that has happened here, especially with the coroutines.

Please create a separate component with a descriptive name, and put all logics related to decreasing the health there. This way, you actually don't even need to use coroutines.

Don't get me wrong; you did a great job, and I'm saying this because coroutines are hard to grasp. I can definitely see you put some effort to understand the code that was written by someone else, and I know that is not an easy thing to do. The suggestions I am giving you here are not for critisizing you, but rather for sharing my thoughts on and styles of programming.

Also, please feel free to ask questions or request explanations while you are on your task! Game development is a huge thing, and it is really really hard to figure out what exact steps to take to achieve what you want. I think some guidance on where to look at and what resources are useful can be a great help, if I'm not bad at explaining things of course...

To recap, make a new MonoBehaviour and move your logics to the new component. If coroutines are too difficult to understand, you can just use plain `Start` and `Update` event functions to accomplish the task. Thanks for your work!

File: Assets/Scripts/PlayerCharacter/PlayerHPDecreasement.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+﻿using Combat;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[RequireComponent(typeof(CombatEntity))]
+public class PlayerHPDecreasement : MonoBehaviour
+{
+    [SerializeField] public int hpDecreasingValue = 1;
```
Comment (JameeKim) [inline Assets/Scripts/PlayerCharacter/PlayerHPDecreasement.cs]: You do not need this `SerializeField` attribute on public fiels that are serializable. Unity serializes those fields automatically, and that is why you can see the values in the inspector.

File: Assets/Scripts/Combat/CombatEntity.cs [hunk 5]
```diff
@@ -122,9 +122,9 @@ public void GetAttacked(int rawDamage)
             InternalCurrentHealth = currentHealth - (rawDamage - currentDefense);
         }
 
-        public void GetDecreased(int rawDamage)
+        public void DecreasePlayerHealth(int hpDecreasingValue)
```
Comment (JameeKim) [inline Assets/Scripts/Combat/CombatEntity.cs]: Nice details! But maybe got too specific. This component is `CombatEntity`, so there is no guarantee that the entity will be the player character. We are adding this method because we are trying to decrease the player health, but who knows if we'll need it later on for some other purposes? `DecreaseHealth` would be enough. :3

**Review by JameeKim — CHANGES_REQUESTED** [review]
Thank you, you've done a great work! There are just a couple of little things I pointed out. We can definitely merge this after those tiny changes!

**Review by JameeKim — APPROVED** [review]
Nice work! Thank you!
